File: graphz/src/main/scala/coop/rchain/graphz/Graphz.scala
package coop.rchain.graphz

import java.io.FileOutputStream
import cats._
import cats.effect.{Concurrent, Sync}
import cats.effect.concurrent.Ref
import cats.syntax.all._

trait GraphSerializer[F[_]] {
  def push(str: String, suffix: String = "\n"): F[Unit]
}

class StringSerializer[F[_]: Concurrent](ref: Ref[F, StringBuffer]) extends GraphSerializer[F] {
  override def push(str: String, suffix: String): F[Unit] = ref.update(_.append(str + suffix))
}

class ListSerializer[F[_]: Concurrent](ref: Ref[F, Vector[String]]) extends GraphSerializer[F] {
  override def push(str: String, suffix: String): F[Unit] = ref.update(_ :+ (str + suffix))
}

class FileSerializer[F[_]: Sync](fos: FileOutputStream) extends GraphSerializer[F] {
  override def push(str: String, suffix: String): F[Unit] = Sync[F].delay {
    fos.write(str.getBytes)
    fos.flush()
  }
}

sealed trait GraphType
case object Graph   extends GraphType
case object DiGraph extends GraphType

sealed trait GraphShape
case object Circle       extends GraphShape
case object DoubleCircle extends GraphShape
case object Box          extends GraphShape
case object PlainText    extends GraphShape
case object Msquare      extends GraphShape
case object Record       extends GraphShape

sealed trait GraphRank
case object Same   extends GraphRank
case object Min    extends GraphRank
case object Source extends GraphRank
case object Max    extends GraphRank
case object Sink   extends GraphRank

sealed trait GraphRankDir
case object TB extends GraphRankDir
case object BT extends GraphRankDir
case object LR extends GraphRankDir
case object RL extends GraphRankDir

sealed trait GraphStyle
case object Solid  extends GraphStyle
case object Bold   extends GraphStyle
case object Filled extends GraphStyle
case object Invis  extends GraphStyle
case object Dotted extends GraphStyle
case object Dashed extends GraphStyle

sealed trait GraphArrowType
case object NormalArrow extends GraphArrowType
case object InvArrow    extends GraphArrowType
case object NoneArrow   extends GraphArrowType

object Graphz {

  implicit val showShape: Show[GraphShape] = new Show[GraphShape] {
    def show(shape: GraphShape): String = shape match {
      case Circle       => "circle"
      case DoubleCircle => "doublecircle"
      case Box          => "box"
      case PlainText    => "plaintext"
      case Msquare      => "Msquare"
      case Record       => "record"
    }
  }

  def smallToString[A]: Show[A] = (a: A) => a.toString.toLowerCase

  implicit val showStyle: Show[GraphStyle]     = smallToString[GraphStyle]
  implicit val showRank: Show[GraphRank]       = smallToString[GraphRank]
  implicit val showRankDir: Show[GraphRankDir] = Show.fromToString[GraphRankDir]
  implicit val showArrowType: Show[GraphArrowType] = new Show[GraphArrowType] {
    def show(arrowType: GraphArrowType): String = arrowType match {
      case NormalArrow => "normal"
      case InvArrow    => "inv"
      case NoneArrow   => "none"
    }
  }

  def DefaultShape: GraphShape = Circle

  def apply[F[_]: Monad](
      name: String,
      gtype: GraphType,
      ser: GraphSerializer[F],
      subgraph: Boolean = false,
      comment: Option[String] = None,
      label: Option[String] = None,
      splines: Option[String] = None,
      rank: Option[GraphRank] = None,
      rankdir: Option[GraphRankDir] = None,
      style: Option[String] = None,
      color: Option[String] = None,
      node: Map[String, String] = Map.empty
  ): F[Graphz[F]] = {

    def insert(str: Option[String], v: String => String): F[Unit] = {
      val indent = if (subgraph) tab + tab else tab
      str.fold(().pure[F])(s => ser.push(indent + v(s)))
    }

    for {
      _ <- comment.fold(().pure[F])(c => ser.push(s"// $c"))
      t = if (subgraph) s"$tab$tab" else tab
      _ <- ser.push(head(gtype, subgraph, name))
      _ <- insert(label, l => s"label = ${quote(l)}")
      _ <- insert(style, s => s"style=$s")
      _ <- insert(color, s => s"color=$s")
      _ <- insert(rank.map(_.show), r => s"rank=$r")
      _ <- insert(rankdir.map(_.show), r => s"rankdir=$r")
      _ <- insert(attrMkStr(node), n => s"node $n")
      _ <- insert(splines.map(_.show), s => s"splines=$s")
    } yield new Graphz[F](gtype, t, ser)
  }

  def subgraph[F[_]: Monad](
      name: String,
      gtype: GraphType,
      ser: GraphSerializer[F],
      label: Option[String] = None,
      rank: Option[GraphRank] = None,
      rankdir: Option[GraphRankDir] = None,
      style: Option[String] = None,
      color: Option[String] = None
  ): F[Graphz[F]] =
    apply[F](
      name,
      gtype,
      subgraph = true,
      label = label,
      rank = rank,
      rankdir = rankdir,
      style = style,
      color = color,
      ser = ser
    )

  private def head(gtype: GraphType, subgraph: Boolean, name: String): String = {
    val prefix = (gtype, subgraph) match {
      case (_, true)    => s"${tab}subgraph"
      case (Graph, _)   => s"graph"
      case (DiGraph, _) => s"digraph"
    }
    if (name == "") s"$prefix {" else s"""$prefix "$name" {"""
  }

  def quote(str: String): String = str match {
    case _ if str.startsWith("\"") => str
    case _                         => s""""$str""""
  }

  def attrMkStr(attr: Map[String, String]): Option[String] =
    if (attr.isEmpty) None
    else
      Some("[" + attr.map(t => t._1 + "=" + t._2).mkString(" ") + "]")

  val tab = "  "
}

class Graphz[F[_]: Monad](gtype: GraphType, t: String, val ser: GraphSerializer[F]) {

  def edge(edg: (String, String)): F[Unit] = edge(edg._1, edg._2)
  def edge(
      src: String,
      dst: String,
      style: Option[GraphStyle] = None,
      arrowHead: Option[GraphArrowType] = None,
      constraint: Option[Boolean] = None
  ): F[Unit] = {
    import Graphz.{showArrowType, showStyle}
    val attrStyle: Map[String, String] = style.map(s => Map("style" -> s.show)).getOrElse(Map.empty)
    val attrArrowHead: Map[String, String] =
      arrowHead.map(s => Map("arrowhead" -> s.show)).getOrElse(Map.empty)
    val attrConstraint: Map[String, String] =
      constraint.map(s => Map("constraint" -> s.show)).getOrElse(Map.empty)
    val attrs: Map[String, String] = attrStyle |+| attrConstraint |+| attrArrowHead
    ser.push(
      edgeMkStr.format(
        Graphz.quote(src),
        Graphz.quote(dst),
        Graphz.attrMkStr(attrs).map(a => " " + a).getOrElse("")
      )
    )
  }

  def node(
      name: String,
      shape: GraphShape = Circle,
      style: Option[GraphStyle] = None,
      color: Option[String] = None,
      label: Option[String] = None
  ): F[Unit] = {
    import Graphz.{showShape, showStyle}
    val attrShape: Map[String, String] =
      if (shape == Graphz.DefaultShape) Map.empty else Map("shape" -> shape.show)
    val attrStyle: Map[String, String] = style.map(s => Map("style" -> s.show)).getOrElse(Map.empty)
    val attrColor: Map[String, String] = color.map(c => Map("color" -> c)).getOrElse(Map.empty)
    val attrLabel: Map[String, String] = label.map(c => Map("label" -> c)).getOrElse(Map.empty)

    val attrs: Map[String, String] = attrShape |+| attrColor |+| attrLabel |+| attrStyle
    ser.push(t + Graphz.quote(name) + Graphz.attrMkStr(attrs).map(a => " " + a).getOrElse(""))
  }

  def close: F[Unit] = {
    val content = t.substring(Graphz.tab.length)
    val suffix  = if (content.isEmpty) "" else "\n"
    ser.push(s"$content}", suffix = suffix)
  }

  private def edgeMkStr: String = gtype match {
    case Graph   => s"$t%s -- %s%s"
    case DiGraph => s"$t%s -> %s%s"
  }
}

File: graphz/src/rust/graphz.rs
use async_trait::async_trait;
use std::collections::HashMap;
use std::fmt::{self, Display};
use std::sync::Arc;
use tokio::fs::File;
use tokio::io::AsyncWriteExt;
use tokio::sync::Mutex;

#[async_trait]
pub trait GraphSerializer: Send + Sync {
    async fn push(&self, str: &str, suffix: &str) -> Result<(), GraphzError>;
}

pub struct StringSerializer {
    buffer: Arc<Mutex<String>>,
}

impl StringSerializer {
    pub fn new() -> Self {
        Self {
            buffer: Arc::new(Mutex::new(String::new())),
        }
    }

    pub async fn get_content(&self) -> String {
        self.buffer.lock().await.clone()
    }
}

impl Default for StringSerializer {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl GraphSerializer for StringSerializer {
    async fn push(&self, str: &str, suffix: &str) -> Result<(), GraphzError> {
        let mut buffer = self.buffer.lock().await;
        buffer.push_str(str);
        buffer.push_str(suffix);
        Ok(())
    }
}

pub struct ListSerializer {
    lines: Arc<Mutex<Vec<String>>>,
}

impl ListSerializer {
    pub fn new() -> Self {
        Self {
            lines: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub async fn get_lines(&self) -> Vec<String> {
        self.lines.lock().await.clone()
    }
}

impl Default for ListSerializer {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl GraphSerializer for ListSerializer {
    async fn push(&self, str: &str, suffix: &str) -> Result<(), GraphzError> {
        let mut lines = self.lines.lock().await;
        lines.push(format!("{}{}", str, suffix));
        Ok(())
    }
}

pub struct FileSerializer {
    file: Arc<Mutex<File>>,
}

impl FileSerializer {
    pub async fn new(file: File) -> Self {
        Self {
            file: Arc::new(Mutex::new(file)),
        }
    }
}

#[async_trait]
impl GraphSerializer for FileSerializer {
    async fn push(&self, str: &str, suffix: &str) -> Result<(), GraphzError> {
        let mut file = self.file.lock().await;
        let content = format!("{}{}", str, suffix);
        file.write_all(content.as_bytes())
            .await
            .map_err(|e| GraphzError::IoError(e.to_string()))?;
        file.flush()
            .await
            .map_err(|e| GraphzError::IoError(e.to_string()))?;
        Ok(())
    }
}

// Sealed traits equivalent - all enum definitions first (like Scala sealed traits)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphType {
    Graph,
    DiGraph,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphShape {
    Circle,
    DoubleCircle,
    Box,
    PlainText,
    Msquare,
    Record,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphRank {
    Same,
    Min,
    Source,
    Max,
    Sink,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphRankDir {
    TB,
    BT,
    LR,
    RL,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphStyle {
    Solid,
    Bold,
    Filled,
    Invis,
    Dotted,
    Dashed,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphArrowType {
    NormalArrow,
    InvArrow,
    NoneArrow,
}

// Display implementations (equivalent to Scala's Show instances in object Graphz)

// Note: We don't use a small_to_string function like Scala's smallToString[A]
// because:
// 1. It would depend on Debug implementation which can change
// 2. It would require runtime string allocation + processing
// 3. Explicit match provides compile-time safety and zero-cost abstractions
// 4. Rust doesn't have Scala's type-level abstractions for Show[A]

impl Display for GraphShape {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            GraphShape::Circle => "circle",
            GraphShape::DoubleCircle => "doublecircle",
            GraphShape::Box => "box",
            GraphShape::PlainText => "plaintext",
            GraphShape::Msquare => "Msquare",
            GraphShape::Record => "record",
        };
        write!(f, "{}", s)
    }
}

impl Display for GraphStyle {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            GraphStyle::Solid => "solid",
            GraphStyle::Bold => "bold",
            GraphStyle::Filled => "filled",
            GraphStyle::Invis => "invis",
            GraphStyle::Dotted => "dotted",
            GraphStyle::Dashed => "dashed",
        };
        write!(f, "{}", s)
    }
}

impl Display for GraphRank {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            GraphRank::Same => "same",
            GraphRank::Min => "min",
            GraphRank::Source => "source",
            GraphRank::Max => "max",
            GraphRank::Sink => "sink",
        };
        write!(f, "{}", s)
    }
}

impl Display for GraphRankDir {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

impl Display for GraphArrowType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            GraphArrowType::NormalArrow => "normal",
            GraphArrowType::InvArrow => "inv",
            GraphArrowType::NoneArrow => "none",
        };
        write!(f, "{}", s)
    }
}

// Public functions corresponding to Scala's object Graphz
const TAB: &str = "  ";

pub fn default_shape() -> GraphShape {
    GraphShape::Circle
}

pub async fn apply(
    name: String,
    gtype: GraphType,
    ser: Arc<dyn GraphSerializer>,
    subgraph: bool,
    comment: Option<String>,
    label: Option<String>,
    splines: Option<String>,
    rank: Option<GraphRank>,
    rankdir: Option<GraphRankDir>,
    style: Option<String>,
    color: Option<String>,
    node: HashMap<String, String>,
) -> Result<Graphz, GraphzError> {
    let insert = |opt_str: Option<String>, transform: fn(&str) -> String| {
        let ser = ser.clone();
        async move {
            let indent = if subgraph {
                format!("{}{}", TAB, TAB)
            } else {
                TAB.to_string()
            };

            if let Some(s) = opt_str {
                ser.push(&format!("{}{}", indent, transform(&s)), "\n")
                    .await?;
            }
            Ok::<(), GraphzError>(())
        }
    };

    if let Some(c) = &comment {
        ser.push(&format!("// {}", c), "\n").await?;
    }

    let t = if subgraph {
        format!("{}{}", TAB, TAB)
    } else {
        TAB.to_string()
    };

    let header = head(gtype, subgraph, &name);
    ser.push(&header, "\n").await?;

    insert(label, |l| format!("label = {}", quote(l))).await?;
    insert(style, |s| format!("style={}", s)).await?;
    insert(color, |s| format!("color={}", s)).await?;
    insert(rank.map(|r| r.to_string()), |r| format!("rank={}", r)).await?;
    insert(rankdir.map(|r| r.to_string()), |r| format!("rankdir={}", r)).await?;
    insert(attr_mk_str(&node), |n| format!("node {}", n)).await?;
    insert(splines, |s| format!("splines={}", s)).await?;

    Ok(Graphz {
        graph_type: gtype,
        tab: t,
        serializer: ser,
    })
}

pub async fn subgraph(
    name: String,
    gtype: GraphType,
    ser: Arc<dyn GraphSerializer>,
    label: Option<String>,
    rank: Option<GraphRank>,
    rankdir: Option<GraphRankDir>,
    style: Option<String>,
    color: Option<String>,
) -> Result<Graphz, GraphzError> {
    apply(
        name,
        gtype,
        ser,
        true,
        None,
        label,
        None,
        rank,
        rankdir,
        style,
        color,
        HashMap::new(),
    )
    .await
}

fn head(graph_type: GraphType, subgraph: bool, name: &str) -> String {
    let prefix = match (graph_type, subgraph) {
        (_, true) => format!("{}subgraph", TAB),
        (GraphType::Graph, _) => "graph".to_string(),
        (GraphType::DiGraph, _) => "digraph".to_string(),
    };

    if name.is_empty() {
        format!("{} {{", prefix)
    } else {
        format!("{} \"{}\" {{", prefix, name)
    }
}

fn quote(s: &str) -> String {
    if s.starts_with('"') {
        s.to_string()
    } else {
        format!("\"{}\"", s)
    }
}

fn attr_mk_str(attrs: &HashMap<String, String>) -> Option<String> {
    if attrs.is_empty() {
        None
    } else {
        let attr_pairs: Vec<String> = attrs.iter().map(|(k, v)| format!("{}={}", k, v)).collect();
        Some(format!("[{}]", attr_pairs.join(" ")))
    }
}

pub struct Graphz {
    graph_type: GraphType,
    tab: String,
    serializer: Arc<dyn GraphSerializer>,
}

impl Graphz {
    pub async fn edge_tuple(&self, edg: (String, String)) -> Result<(), GraphzError> {
        self.edge(&edg.0, &edg.1, None, None, None).await
    }

    pub async fn edge(
        &self,
        src: &str,
        dst: &str,
        style: Option<GraphStyle>,
        arrow_head: Option<GraphArrowType>,
        constraint: Option<bool>,
    ) -> Result<(), GraphzError> {
        let mut attrs = HashMap::new();
        if let Some(s) = style {
            attrs.insert("style".to_string(), s.to_string());
        }

        if let Some(ah) = arrow_head {
            attrs.insert("arrowhead".to_string(), ah.to_string());
        }

        if let Some(c) = constraint {
            attrs.insert("constraint".to_string(), c.to_string());
        }

        let quoted_src = quote(src);
        let quoted_dst = quote(dst);
        let attr_str = attr_mk_str(&attrs)
            .map(|a| format!(" {}", a))
            .unwrap_or_default();

        let formatted_edge = self.edge_mk_str(&quoted_src, &quoted_dst, &attr_str);
        self.serializer.push(&formatted_edge, "\n").await
    }

    pub async fn node(
        &self,
        name: &str,
        shape: GraphShape,
        style: Option<GraphStyle>,
        color: Option<String>,
        label: Option<String>,
    ) -> Result<(), GraphzError> {
        let mut attrs = HashMap::new();

        if shape != default_shape() {
            attrs.insert("shape".to_string(), shape.to_string());
        }

        if let Some(s) = style {
            attrs.insert("style".to_string(), s.to_string());
        }

        if let Some(c) = color {
            attrs.insert("color".to_string(), c);
        }

        if let Some(l) = label {
            attrs.insert("label".to_string(), l);
        }

        let quoted_name = quote(name);
        let attr_str = attr_mk_str(&attrs)
            .map(|a| format!(" {}", a))
            .unwrap_or_default();

        let node_str = format!("{}{}{}", self.tab, quoted_name, attr_str);
        self.serializer.push(&node_str, "\n").await
    }

    pub async fn close(&self) -> Result<(), GraphzError> {
        let content = &self.tab[TAB.len()..];

        // Equivalent to Scala: val suffix = if (content.isEmpty) "" else "\n"
        let suffix = if content.is_empty() { "" } else { "\n" };

        self.serializer
            .push(&format!("{}}}", content), suffix)
            .await
    }

    fn edge_mk_str(&self, src: &str, dst: &str, attr_str: &str) -> String {
        match self.graph_type {
            GraphType::Graph => format!("{}{} -- {}{}", self.tab, src, dst, attr_str),
            GraphType::DiGraph => format!("{}{} -> {}{}", self.tab, src, dst, attr_str),
        }
    }
}

// Error type for graphz operations
#[derive(Debug, Clone)]
pub enum GraphzError {
    IoError(String),
    SerializationError(String),
}

impl Display for GraphzError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GraphzError::IoError(msg) => write!(f, "IO Error: {}", msg),
            GraphzError::SerializationError(msg) => write!(f, "Serialization Error: {}", msg),
        }
    }
}

impl std::error::Error for GraphzError {}
